src/utils/sources-tree/addToTree.js

```diff
@@ -84,18 +204,24 @@ function addSourceToNode(node: Node, url: Object, source: SourceRecord) {
   }
 
   const name = getFilenameFromPath(url.path);
-  const existingNode = node.contents.find(childNode => childNode.name === name);
+  const childIndex = findNodeInContents(
+    node,
+    createTreeNodeMatcher(name, false, null)
+  );
 
   // if we are readding an existing file in the node, overwrite the existing
   // file and return the node's contents
-  if (existingNode) {
+  if (childIndex >= 0) {
+    const existingNode = node.contents[childIndex];
     existingNode.contents = source;
     return node.contents;
   }
 
   // if this is a new file, add the new file;
   const newNode = createNode(name, source.get("url"), source);
-  return [...node.contents, newNode];
+  const contents = node.contents.slice(0);
+  contents.splice(-childIndex - 1, 0, newNode);
```

Comment:
it is so unclear what is going on here. we shouldn't be using negative numbers to determine if something is a new node

Reply:
Changed return type to return an object (I hope JIT will not create lots of objects in memory to GC them later). Measured performance for 3k file -- there was no real change in the performance.